parlai/core/torch_ranker_agent.py

```diff
@@ -0,0 +1,351 @@
+            loss = self.rank_loss(scores, label_inds)
+            self.metrics['loss'] += loss.item()
+            self.metrics['examples'] += batchsize
+            for b in range(batchsize):
+                rank = (ranks[b] == label_inds[b]).nonzero().item()
+                self.metrics['rank'] += 1 + rank
+
+        cand_preds = [[cands[i] for i in row] for row in ranks]
+        preds = [row[0] for row in cand_preds]
+        return Output(preds, cand_preds)
+
+    def _build_candidates(self, batch, source, mode):
+        """Build a candidate set for this batch
+
+        :param batch: a Batch object (defined in torch_agent.py)
+        :param source: the source from which candidates should be built, one of
+            ['batch', 'inline', 'fixed', 'vocab']
+        :param mode: 'train' or 'eval'
+
+        :return: tuple of tensors (label_inds, cands, cand_vecs)
+            label_inds: A [bsz] LongTensor of the indices of the labelf or each
+                example in the tensor of candidates
+            cands: A [num_cands] list of (text) candidates
+            cand_vecs: A padded [num_cands, seqlen] LongTensor of vectorized
+                candidates
+
+        Possible sources of candidates:
+            * batch: the set of all labels in this batch
+                Use all labels in the batch as the candidate set (with all but
+                the example's label being treated as negatives). If labels
+                are single tokens, filter to unique labels. (Ideally, we would
+                always filter to unique labels, but we abstain for the sake of
+                speed.)
+            * inline: batch_size lists, one list per example
+                If each example comes with a list of possible candidates, use
+                those.
+            * fixed: one global candidate list, provide in a file from the user
+                If self.fixed_candidates is not None, use a set of fixed
+                candidates for all examples.
+                Note: this setting is not recommended for training unless the
+                universe of possible candidates is very small.
+            * vocab: the set of all tokens in the vocabulary
+
+        TODO: Consider allowing examples in a batch to have different candidate
+        sets (requires changes to MemNN module). Currently, all examples in a
+        batch must have the same candidate set.
+        """
+        label_vecs = batch.label_vec # [bsz] list of lists of LongTensors
+        batchsize = batch.text_vec.shape[0]
+
+        if label_vecs is not None:
+            assert label_vecs.dim() == 2
+        label_inds = None
+
+        if source == 'batch':
+            self._warn_once(
+                f'{mode}_batch_candidates',
+                f'[ Executing {mode} mode with batch labels as set of '
+                    'candidates. ]')
+            if batchsize == 1:
+                self._warn_once(
+                    f'{mode}_batchsize_1',
```

Comment:
Need to get read of the fstrings. We're currently maintaining 3.5 compatibility. @alexholdenmiller might take the excuse to drop it now though.

Reply:
Yes please, @alexholdenmiller! f-strings are so much more readable and take up less space. 🙏